src/simulation_engine.cpp

```diff
@@ -1,21 +1,92 @@
 #include <iostream>
 #include <vector>
 #include <memory>
+#include <random>
+#include <algorithm>
+#include <iterator>
 
 #include <simulator/simulation_engine.hpp>
 #include <simulator/simulation_map.hpp>
+#include <simulator/bug.hpp>
 
 namespace simulation {
 
-SimulationEngine::SimulationEngine() {}
+const uint16_t ANTS = 10;
+const uint16_t DOODLEBUGS = 3;
 
-void SimulationEngine::simulate_round() {}
+SimulationEngine::SimulationEngine(): map_(std::make_shared<SimulationMap>()){}
 
-void SimulationEngine::populate_map_with_species() {}
+void SimulationEngine::start_simulation(uint16_t n_rounds) {
+  for(;n_rounds && simulate_round();n_rounds--);
+}
 
-void SimulationEngine::on_bug_born(std::shared_ptr<Bug>) {}
+bool SimulationEngine::simulate_round() {
+  uint16_t species_left = 0;
+  std::cout << "--------" << std::endl;
+  std::cout << "[SimulationEngine] New Round" << std::endl;
+    map_->print_map();
   
-void SimulationEngine::on_bug_died(std::shared_ptr<Bug>) {}
+  // Iterates over cells and makes every bug move/breed/eat/starve accordingly.
+  for(auto cell: map_->get_cells()) {
+    auto bug_ref = cell->get_bug();
+    if(bug_ref && !bug_ref->has_played()) {
+      bug_ref->play_round();
+    }
+  }
+
+  // Analyzes results of the rounds.
+  bool only_ants = true;
+  bool only_doo = true;
+  for(auto cell: map_->get_cells()) {
+    if(cell->get_bug()) {
+      species_left++;
+      only_doo &= !cell->get_bug()->is_prey();
+      only_ants &= cell->get_bug()->is_prey();
```

Comment:
Again, I don't like checking if it's a prey to see if it's an ant. You should had been more explicit

Reply:
Oh, I thought we were aware that this was a weak spot of the code if we eventually wanted to scale it, but didn't know it was a requirement to change it. 
If that's the case, I'd do so in a separate PR though, It's a mid-sized change, I don't want to increase the size of this PR that is already too big.

Reply:
not required to change it. I was aware that we agreed to leave it like. Just stressing it again because I'm tough :stuck_out_tongue_closed_eyes: